File: termite/termite1.py
# -*- coding: utf-8 -*-
"""
Éditeur de Spyder

Ceci est un script temporaire.
"""
import numpy as np
import matplotlib.pyplot as plt
import pdb

from pytictoc import TicToc
t = TicToc() #create instance of class


#plt.scatter(x, y, s=area, c=colors, alpha = 0.5)

posXmax = 600
posYmax = 600

speedX = 6
speedY = 6

tSize = 2.5

class Termite:
    
    def __init__(self, posX, posY, posXmax, posYmax, tSize):
        self.positionX = posX
        self.positionY = posY
        self.tSize = tSize
        self.color = "#1A2930"
        self.carryChip = 0
        self.posXmax = posXmax
        self.posYmax = posYmax
        self.boundsLeft = self.positionX - 1*self.tSize
        self.boundsRight = self.positionX + 1*self.tSize
        self.boundsBottom = self.positionY - 1*self.tSize
        self.boundsTop = self.positionY + 1*self.tSize
        self.plotArea = np.pi * self.tSize ** 2
          
    def move(self, speedX, speedY):
        self.positionX += (np.random.random() * 2 - 1) * speedX
        self.positionY += (np.random.random() * 2 - 1) * speedY
        
        self.boundsLeft = self.positionX - 1*self.tSize
        self.boundsRight = self.positionX + 1*self.tSize
        self.boundsBottom = self.positionY - 1*self.tSize
        self.boundsTop = self.positionY + 1*self.tSize
     
        
        if (self.boundsLeft > self.posXmax):
            self.positionX = 0
        
        if (self.boundsRight < 0):
            self.positionX = self.posXmax
            
        if (self.boundsTop > self.posYmax):
            self.positionY = 0
            
        if (self.boundsBottom < 0):
            self.positionY = self.posYmax



class Chip:
    
    def __init__(self,xpos, ypos, radius):
        self.radius = radius
        self.color = "#FF533D"
        self.positionX = xpos
        self.positionY = ypos
        self.plotArea = np.pi * self.radius ** 2


def scatterPlot(tGroup, wGroup):    
    tposX = [termite.positionX for termite in tGroup]
    tposY = [termite.positionY for termite in tGroup]    
    tColor = [termite.color for termite in tGroup] 
    tArea =  [termite.plotArea for termite in tGroup]    

    cposX = [c.positionX for c in wGroup]
    cposY = [c.positionY for c in wGroup]    
    cColor = [c.color for c in wGroup] 
    cArea =  [c.plotArea for c in wGroup] 

    AposX = cposX + tposX
    AposY = cposY + tposY
    AColor = cColor + tColor
    AArea = cArea + tArea

    fig = plt.figure(facecolor = "#F5F5F5")

    ax=fig.add_axes([0,0,1,1])
    ax.set_axis_off()                 
    ax.scatter(AposX, AposY, AArea, AColor)
    #plt.axis("off")
    ax.set_xlim([0,posXmax])
    ax.set_ylim([0,posYmax])

    ax.set_aspect('auto')
    plt.show()
    

nTermites = 100

tGroup = []

#populate Termite list
for i in range(1,nTermites+1):
    posX = np.random.random() * posXmax
    posY = np.random.random() * posYmax
    tGroup.append(Termite(posX, posY, posXmax, posYmax, tSize))

#populate Wood chips list

nChips = 2000

cRadius = 1
wGroup = []

for j in range(0 , nChips+1):
    
    wGroup.append(Chip(np.random.random() * posXmax,np.random.random() * posYmax,cRadius))


#scatterPlot(tGroup, wGroup)


t.tic()
#input("Press enter to continue...")
nIter = 1000

for n in range(1,nIter+1):
    if ( n % 100 == 0):
        print(n)
    for i in range(0,nTermites):
        vAnt = tGroup[i]
        
        vAnt.move(speedX, speedY)
        
        for j in range(0, len(wGroup)-1):
            
            wc = wGroup[j]
            dist2X = ( vAnt.positionX - wc.positionX ) ** 2
            dist2Y = ( vAnt.positionY - wc.positionY ) ** 2
    
            if (dist2X + dist2Y <= vAnt.tSize ** 2):
                if (vAnt.carryChip):
                    vAnt.carryChip = 0
                    wGroup.append(Chip(vAnt.positionX, vAnt.positionY, cRadius))
                    vAnt.color = "#1A2930"
                    vAnt.positionX += 10* np.random.random() -5
                    vAnt.positionY += 10* np.random.random() -5
                else:
                    vAnt.carryChip = 1
                    vAnt.color = "green"
                    wGroup.remove(wc)
                    #print(len(wGroup))
                    #pdb.set_trace()
                    

t.toc()
print("after " + str(nIter) + " Iterations")
scatterPlot(tGroup, wGroup)  

File: termite/termite1.jl
using PyPlot

posXmax = 600.0
posYmax = 600.0

tSize = 2.5
cRadius = 1

mutable struct Position
    x::Float64
    y::Float64
end

mutable struct BoundingBox
    left::Float64
    right::Float64
    top::Float64
    bottom::Float64
end

BoundingBox(pos::Position, size::Float64) = BoundingBox(pos.x-size, pos.x+size, pos.y + size, pos.y-size)

mutable struct ChipCarrier
    color::String
    hasChip::Bool
    ChipCarrier(hasChip) = hasChip ? new("green", hasChip) : new("#1A2930", hasChip)
end

struct Termite
    pos::Position
    bounds::BoundingBox
    carriesChip::ChipCarrier
    plotArea::Float64
end

Termite(pos::Position) = Termite(pos,
                    BoundingBox(pos,tSize),
                    ChipCarrier(false),
                    pi * tSize ^ 2
                    )


struct Chip
    pos::Position
    color::String
    radius::Float64
    plotArea::Float64

end

Chip(pos::Position) = Chip(pos, "#FF533D", cRadius, pi * cRadius ^ 2)

function updateBoundingBox(t::Termite)
    t.bounds.left = t.pos.x - tSize;
    t.bounds.right = t.pos.x + tSize;
    t.bounds.bottom = t.pos.y - tSize;
    t.bounds.top = t.pos.y + tSize;

end

move = function(t::Termite, speedX::Float64, speedY::Float64)
    t.pos.x += (2 * rand() - 1) * speedX;
    t.pos.y += (2 * rand() - 1) * speedY;
    updateBoundingBox(t)

    if (t.bounds.left > posXmax)
        t.pos.x = 0
        updateBoundingBox(t)
    end

    if (t.bounds.right < 0)
        t.pos.x = posXmax
        updateBoundingBox(t)
    end

    if (t.bounds.bottom > posYmax)
        t.pos.y = 0
        updateBoundingBox(t)
    end

    if (t.bounds.top < 0)
        t.pos.y = posYmax
        updateBoundingBox(t)
    end
end


t = Termite(Position(3.0,3.0))

nTermites = 100;

tGroup = Array{Termite}(nTermites);

for i=1:nTermites
    #add termite to array
    tGroup[i] = Termite(Position(posXmax * rand(), posYmax * rand()))
end

nChips = 2000;

#wGroup = Array{Chip}();
wGroup = (Chip)[];
#now the pain: plotting
for i=1:nChips
    push!(wGroup, Chip(Position(posXmax * rand(), posYmax * rand())));
end

function scatterPlot(tGroup, wGroup)

cposX = [ce.pos.x for ce in wGroup]
cposY = [ce.pos.y for ce in wGroup]
cColor = [ce.color for ce in wGroup]
cArea = [ce.plotArea for ce in wGroup]


tposX = [te.pos.x for te in tGroup]
tposY = [te.pos.y for te in tGroup]
tColor = [te.carriesChip.color for te in tGroup]
tArea = [te.plotArea for te in tGroup]

AposX = vcat(cposX, tposX)
AposY = vcat(cposY,tposY)
AColor = vcat(cColor, tColor)
AArea = vcat(cArea, tArea)



# x = linspace(0,2*pi,1000); y = sin.(3 * x + 4 * cos.(2 * x));
# plot(x, y, color="red", linewidth=2.0, linestyle="--")
# title("A sinusoidally modulated sinusoid")
fig = figure(facecolor="#F5F5F5")
axis([0,posXmax, 0, posYmax])
scatter(AposX, AposY, AArea, AColor)
#scatter(0,0,5,"black")
axis("off")
show()
#savefig("termite.png", facecolor="#F5F5F5", edgecolor="none", bbox_inches="tight", pad_inches=0)
end


scatterPlot(tGroup, wGroup)

#wc = wGroup[100]
#filter!(x -> x!= wc, wGroup)

nIter = 400

speedX = 6.0
speedY = 6.0

lwg = length(wGroup)

tic()

for n in 1:nIter
    if (n % 100 == 0)
        println(n)
    end
    for i in 1:nTermites
        vAnt = tGroup[i]
        move(vAnt, speedX, speedY)
        #println(length(wGroup))
        for j in 1:length(wGroup)
            wc = wGroup[j]
            #wc = wGroup[j]
            dist2X = (vAnt.pos.x - wc.pos.x) ^ 2
            dist2Y = (vAnt.pos.y - wc.pos.y) ^ 2

            if (dist2X + dist2Y <= tSize ^2)

                if (vAnt.carriesChip.hasChip)
                    vAnt.carriesChip.hasChip = false
                    vAnt.carriesChip.color = "#1A2930"

                    vAnt.pos.x += 10 * rand() -5
                    vAnt.pos.y += 10 * rand() -5
                    push!(wGroup, Chip(Position(vAnt.pos.x, vAnt.pos.y)))

                else
                    vAnt.carriesChip.hasChip = true
                    vAnt.carriesChip.color = "green"
                    filter!(x -> x!= wc, wGroup)
                    #println(length(wGroup))

                end

                break

            end

        end
    end
end

toc()

scatterPlot(tGroup, wGroup)
